File: src/readers/gtfs/realtime/position.ts
import type { Pbf as Protobuf } from '../..';

/** A Position is a point on the Earth's surface. */
export class GTFSRealtimePosition {
  /** Degrees North, in the WGS-84 coordinate system. */
  latitude!: number; // 1 [float]
  /** Degrees East, in the WGS-84 coordinate system. */
  longitude!: number; // 2 [float]
  /**
   * Bearing, in degrees, clockwise from North, i.e., 0 is North and 90 is East.
   * This can be the compass bearing, or the direction towards the next stop
   * or intermediate location.
   * This should not be direction deduced from the sequence of previous
   * positions, which can be computed from previous data.
   */
  bearing?: number; // 3 [float]
  /** Odometer value, in meters. */
  odometer?: number; // 4 [double]
  /** Momentary speed measured by the vehicle, in meters per second. */
  speed?: number; // 5 [float]

  /**
   * @param pbf - The Protobuf object to read from
   * @param end - The end position of the message in the buffer
   */
  constructor(pbf: Protobuf, end: number) {
    pbf.readFields(this.#readTripDescriptor, this, end);
  }

  /**
   * @param tag - The tag of the message
   * @param position - The position to mutate
   * @param pbf - The Protobuf object to read from
   */
  #readTripDescriptor(tag: number, position: GTFSRealtimePosition, pbf: Protobuf) {
    if (tag === 1) position.latitude = pbf.readFloat();
    else if (tag === 2) position.longitude = pbf.readFloat();
    else if (tag === 3) position.bearing = pbf.readFloat();
    else if (tag === 4) position.odometer = pbf.readDouble();
    else if (tag === 5) position.speed = pbf.readFloat();
    else throw new Error(`GTFSRealtimePosition: Unexpected tag: ${tag}`);
  }
}

File: rust/readers/gtfs/realtime/position.rs
use pbf::{ProtoRead, Protobuf};

/// A Position is a point on the Earth's surface.
#[derive(Debug, Default, Clone)]
pub struct GTFSRealtimePosition {
    /// Degrees North, in the WGS-84 coordinate system.
    pub latitude: f32, // 1 [float]
    /// Degrees East, in the WGS-84 coordinate system.
    pub longitude: f32, // 2 [float]
    ///Bearing, in degrees, clockwise from North, i.e., 0 is North and 90 is East.
    ///This can be the compass bearing, or the direction towards the next stop
    ///or intermediate location.
    ///This should not be direction deduced from the sequence of previous
    ///positions, which can be computed from previous data.
    pub bearing: Option<f32>, // 3 [float]
    /// Odometer value, in meters.
    pub odometer: Option<f64>, // 4 [double]
    /// Momentary speed measured by the vehicle, in meters per second.
    pub speed: Option<f32>, // 5 [float]
}
/// Read in the contents of the GTFSRealtimePosition
impl ProtoRead for GTFSRealtimePosition {
    fn read(&mut self, tag: u64, pb: &mut Protobuf) {
        match tag {
            1 => self.latitude = pb.read_varint(),
            2 => self.longitude = pb.read_varint(),
            3 => self.bearing = Some(pb.read_varint()),
            4 => self.odometer = Some(pb.read_varint()),
            5 => self.speed = Some(pb.read_varint()),
            _ => panic!("unknown tag {}", tag),
        }
    }
}
